commit 50aa90ef03007beca2c9108993f5b4f2bb4f0a66
Author: Darrick J. Wong <darrick.wong@oracle.com>
Date:   Mon Jan 8 10:51:04 2018 -0800

    xfs: verify dinode header first

    Move the v3 inode integrity information (crc, owner, metauuid) before we
    look at anything else in the inode so that we don't waste time on a torn
    write or a totally garbled block.  This makes xfs_dinode_verify more
    consistent with the other verifiers.

Signed-off-by: Darrick J. Wong <darrick.wong@oracle.com>
Reviewed-by: Dave Chinner <dchinner@redhat.com>

[ Srinidhi Rao : Backported this fix to 4.9]
Signed-off-by: srinidhira0 <srinidhir@vmware.com>
---
 fs/xfs/libxfs/xfs_inode_buf.c | 23 +++++++++++++----------
 1 file changed, 13 insertions(+), 10 deletions(-)

diff --git a/fs/xfs/libxfs/xfs_inode_buf.c b/fs/xfs/libxfs/xfs_inode_buf.c
index 0c1dd97..5872cd9 100644
--- a/fs/xfs/libxfs/xfs_inode_buf.c
+++ b/fs/xfs/libxfs/xfs_inode_buf.c
@@ -395,6 +395,19 @@ xfs_dinode_verify(
 	if (dip->di_magic != cpu_to_be16(XFS_DINODE_MAGIC))
 		return false;
 
+	/* Verify v3 integrity information first */
+	if (dip->di_version >= 3) {
+		if (!xfs_sb_version_hascrc(&mp->m_sb))
+			return false;
+		if (!xfs_verify_cksum((char *)dip, mp->m_sb.sb_inodesize,
+					XFS_DINODE_CRC_OFF))
+			return false;
+		if (be64_to_cpu(dip->di_ino) != ip->i_ino)
+			return false;
+		if (!uuid_equal(&dip->di_uuid, &mp->m_sb.sb_meta_uuid))
+			return false;
+	}
+
 	/* don't allow invalid i_size */
         di_size = be64_to_cpu(dip->di_size);
         if (di_size & (1ULL << 63))
@@ -474,16 +487,6 @@ xfs_dinode_verify(
 	if (dip->di_version < 3)
 		return true;
 
-	if (!xfs_sb_version_hascrc(&mp->m_sb))
-		return false;
-	if (!xfs_verify_cksum((char *)dip, mp->m_sb.sb_inodesize,
-			      XFS_DINODE_CRC_OFF))
-		return false;
-	if (be64_to_cpu(dip->di_ino) != ip->i_ino)
-		return false;
-	if (!uuid_equal(&dip->di_uuid, &mp->m_sb.sb_meta_uuid))
-		return false;
-
 	flags2 = be64_to_cpu(dip->di_flags2);
 
 	/* don't allow reflink/cowextsize if we don't have reflink */
-- 
2.7.4

